main.go

```diff
@@ -0,0 +1,362 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io/fs"
+	"io/ioutil"
+	"os"
+	"strings"
+)
+
+const (
+	exitFail = 1
+)
+
+//Package level vars to allow patch testing
+type fileLoader func(filename string) ([]byte, error)
+type writeSCP func(filename string, data []byte, perm fs.FileMode) error
+
+var loadFile fileLoader = ioutil.ReadFile
+var saveSCPFile writeSCP = ioutil.WriteFile
+var ErrInvalidParameters = errors.New("input parameters missing")
+var ErrInvalidThreshold = errors.New("threshold limit must be greater than zero")
+var ErrInvalidSCPType = errors.New("scp type must be Allow or Deny")
+
+type SCPRun struct {
+	scannerFilename string
+	serviceType     string
+	serviceName     string
+	thresholdLimit  int64
+	usageData       []byte
+	reports         *[]Report
+	permissionSet   map[string]int64
+	scp             SCP
+}
+
+//validateService checks that the correct apply or
+//deny value was supplied.
+func (s *SCPRun) validateService() (bool, error) {
+	if !checkSCPParameter(s.serviceType) {
+		return false, ErrInvalidSCPType
+	}
+	return true, nil
+}
+
+func (s *SCPRun) getUsageData() error {
+	usageData, err := loadScannerFile(s.scannerFilename)
+	if err != nil {
+		return err
+	}
+	s.usageData = usageData
+	return nil
+}
+
+func (s *SCPRun) getReport() error {
+	r, err := generateReport(s.usageData)
+	if err != nil {
+		return err
+	}
+	s.reports = r
+	return nil
+}
+
+func (s *SCPRun) createPermissions() error {
+	type fnEval = func(int64, int64) bool
```

Comment:
Passing functions as arguments is a nice technique, but it put a bit more pressure on a reader to understand our code. IMHO I would keep it only for a situation where we are not able to use a simple approach.